### PR TITLE
feat(2238): interact mode

### PR DESCRIPTION
**Pull request by wahapo**
## Context
I implemented the interact mode (`sd-local build -i <jobName>`). The implementation makes use of a container that has been set up. ~It's a bit of a stretch to implement, what do you think? We may need to separate the setup part from the launcher if we need to implement it in any other way. If this implementation looks fine, I will also modify the test code.~

## Objective
We will be able to use the `$ sd-local build -i <jobName>` command. `$ exit` to exit run mode.
Although the image shows it runs as `sd-local run <jobName>`, it actually runs as `sd-local build -i <jobName>`.
run

The display of the log output in the launcher is broken. I'd like to display it correctly if possible, but I don't know why.

## References
https://github.com/screwdriver-cd/screwdriver/issues/2238

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by parthasl** [comment]
Looks great @wahapo. Have couple of clarifications

1. runMode => This refers to Interactive mode ?
2. job steps will keep running and runMode shell will open for interactive mode or we have to execute steps individually when runMode is true?
3. will runMode shell exit automatically when job completes or explicitly need to exit?
4. Don't see any impact in launch

**Comment by wahapo** [comment]
@parthasl Thanks for your feedback.

1. Yes, it provides an interactive shell when launched with `sd-local run`. It is for exec any command in a build container started with `sd-local build`. This is before running the steps of screwdriver.yaml.
2. We need to execute steps individually in interactive mode. We will provide a reasonable way to execute the commands of steps in the [STEP2](https://github.com/screwdriver-cd/screwdriver/issues/2238#issuecomment-711872811).
3. Users need to explicitly exec `exit`.
4. If this PR fix looks fine, we don't need to fix the Launcher. The part about saving the environment variables once in a file and exporting them at a different time to recreate the environment is distressing. I was just thinking that if we need to solve this issue, we need to fix the Launcher.

**Comment by tk3fftk** [comment]
it seems `run` and `build` look similar and confusable for users especially new ones. How about `sd-local interact` or `sd-local build -i (--interactive)`?

**Comment by wahapo** [comment]
That may be true, let's talk about the external design in Issue.
https://github.com/screwdriver-cd/screwdriver/issues/2238

**Comment by parthasl** [comment]
Thanks @wahapo everything looks great; only preference is runMode to be changed to interactiveMode

File: launch/docker.go [hunk 1]
```diff
@@ -94,19 +99,41 @@ func (d *docker) runBuild(buildEntry buildEntry) error {
 	artVol := fmt.Sprintf("%s/:%s", hostArtDir, containerArtDir)
 	binVol := fmt.Sprintf("%s:%s", d.volume, "/opt/sd")
 	habVol := fmt.Sprintf("%s:%s", d.habVolume, "/opt/sd/hab")
+
+	// Overwrite steps for sd-local interact mode. The env will load later.
```
Comment (wahapo) [inline launch/docker.go:103]: [note] Discard all the steps defined in screwdriver.yaml.
I would implement a feature to easily call steps in [STEP2](https://github.com/screwdriver-cd/screwdriver/issues/2238#issuecomment-711872811).

File: launch/docker.go [hunk 2]
```diff
@@ -94,19 +99,41 @@ func (d *docker) runBuild(buildEntry buildEntry) error {
...
-	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage, "/opt/sd/local_run.sh", string(configJSON), buildEntry.JobName, environment["SD_API_URL"], environment["SD_STORE_URL"], logfilePath}
+	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage}
+	configJSONArg := string(configJSON)
+	if d.interactMode {
+		configJSONArg = fmt.Sprintf("'%s'", configJSONArg)
+	}
+	launchCommands := []string{"/opt/sd/local_run.sh", configJSONArg, buildEntry.JobName, environment["SD_API_URL"], environment["SD_STORE_URL"], logfilePath}
+	if d.interactMode {
+		dockerCommandOptions = append([]string{"-itd"}, dockerCommandOptions...)
+		dockerCommandOptions = append(dockerCommandOptions, "/bin/sh")
```
Comment (wahapo) [inline launch/docker.go:133]: [note] If it don't run sh in the foreground, it won't be able to attach and interactively run the shell later. We cannot run `local_run.sh` at this stage, we must run it after attaching the build container.

File: launch/interact.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	defer func() { _ = terminal.Restore(int(os.Stdin.Fd()), oldState) }() // Best effort.
+
+	// Start the command
+	c.Start()
+
+	// Copy stdin to the pty and the pty to stdout.
+	go func() {
+		for _, v := range commands {
+			v = append(v, "\n")
+			_, _ = io.Copy(ptmx, strings.NewReader(strings.Join(v, " ")))
```
Comment (wahapo) [inline launch/interact.go:66]: [note] We cann't use `exec.Command` here. `exec.Command` converts the executable command to an absolute path through [LookPath](https://golang.org/pkg/os/exec/#LookPath). The PATH of the execution command is different in the build container than in the environment where sd-local is running.

File: launch/interact.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/creack/pty"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+type Interact interface {
```
Comment (yuichi10) [inline launch/interact.go:17]: How about Interacter?
https://golang.org/doc/effective_go.html#interface-names

File: launch/interact.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	Run(c *exec.Cmd, commands [][]string) error
+}
+
+type InteractImpl struct {
```
Comment (yuichi10) [inline launch/interact.go:21]: How about Interact?

**Comment by yuichi10** [comment]
There are some go lint error. I think it is better to fix theese errors.

File: launch/interact.go (hunk 4 above)
Comment (wahapo) [reply launch/interact.go:17]: Okay.

File: launch/docker.go [hunk 6]
```diff
@@ -94,19 +99,41 @@ func (d *docker) runBuild(buildEntry buildEntry) error {
...
 	if err != nil {
 		return fmt.Errorf("failed to pull user image %v", err)
 	}
 
 	dockerCommandArgs := []string{"container", "run"}
-	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage, "/opt/sd/local_run.sh", string(configJSON), buildEntry.JobName, environment["SD_API_URL"], environment["SD_STORE_URL"], logfilePath}
+	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage}
+	configJSONArg := string(configJSON)
+	if d.interactMode {
+		configJSONArg = fmt.Sprintf("'%s'", configJSONArg)
```
Comment (sugarnaoming) [inline launch/docker.go:128]: I think this work can be done all the time, not just in interactMode.

Comment (wahapo) [reply launch/docker.go]: When running with exec.Command, extra quotations can break user-defined steps if they are enclosed in extra quotes. Somewhere in the escape became incorrect, it waiting for the closing quotation to be typed. It's better to do nothing and let exec.Command do its job.
https://github.com/screwdriver-cd/sd-local/pull/45/files#diff-4f249559a0063721b9902972b3edda755ed28a3110c463e1de09ddb8bd87ecedL138

**Comment by wahapo** [comment]
Added a welcome message and made the input cursor appear without typing enter.

File: cmd/build_test.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ Flags:
   -e, --env stringToString     Set key and value relationship which is set as environment variables of Build Container. (<key>=<value>) (default [])
       --env-file string        Path to config file of environment variables. '.env' format file can be used.
   -h, --help                   help for build
+  -i, --interact               Attach the build container in interactive mode.
```
Comment (jithine) [inline cmd/build_test.go:23]: call this `--interactive`

Comment (wahapo) [reply cmd/build_test.go]: Okay, fixed

**Review by ibu1224 — DISMISSED** [review]

**Review by yoshwata — DISMISSED** [review]

**Review by yuichi10 — APPROVED** [review]
LGTM

File: README.md [hunk 8]
```diff
@@ -45,6 +45,7 @@ Flags:
   -e, --env stringToString     Set key and value relationship which is set as environment variables of Build Container. (<key>=<value>) (default [])
       --env-file string        Path to config file of environment variables. '.env' format file can be used.
   -h, --help                   help for build
+  -i, --interact               Attach the build container in interactive mode.
```
Comment (tk3fftk) [inline README.md:48]:
```suggestion
  -i, --interactive               Attach the build container in interactive mode.
```

File: cmd/build.go [hunk 9]
```diff
@@ -30,6 +30,7 @@ var (
 	osMkdirAll    = os.MkdirAll
 	useSudo       = false
 	usePrivileged = false
+	interactMode  = false
```
Comment (tk3fftk) [inline cmd/build.go:33]: `interactiveMode`?

File: launch/docker.go (hunk 6 above)
Comment (sugarnaoming) [reply launch/docker.go]: I got it.

**Review by jithine — APPROVED** [review]

**Review by kumada626 — APPROVED** [review]

**Comment by wahapo** [comment]
I'll try to merge.

**Pull request merged**